cli/azd/docs/external-prompting.md

```diff
@@ -0,0 +1,114 @@
+      "defaultValue": "<string>", // optional default value, when undefined there is no default.
+    }
+}
+```
+
+The `password` type represents a string value which represents a password. The host may want to use a different UI element (perhaps one that uses `***` instead of characters) when prompting.
+
+The server should respond with 200 OK and the body should be one of the of the following three shapes:
+
+### Success 
+
+When the host is able to prompt the value it is returned with a `result` of `success`:
+
+```jsonc
+{
+    "result": "success", 
+    "value": "<string> | <string>[]"
+}
+```
+
+When the type is `confirm` the value should be either `"true"` or `"false"` (a string, not a JSON boolean) indicating if the user confirmed the operation (`"true"`) or rejected it (`"false"`). Note that a user rejecting a confirm prompt still results in a `"success"` result (the value is simply `"false"`). In the case of `multiSelect` an array of string values is returned, each individual value is a value from the `"options"` array that was selected by the user.
+
+### Cancelled
+
+The user may decline to provide a response (imagine hitting a cancel button on the dialog that is being used to present the question, or a user hitting something like CTRL+C in a terminal interaction to abort the question asking but not the entire application).
+
+In this case the `result` is `cancelled`.
+
+```jsonc
+{
+    "result": "cancelled", 
+}
+```
+
+`azd` returns a special `Error` type internally in this case, which up-stack code can use.
+
+### Error
+
+Some error happened during prompting - the error message will be returned as is.
+
+```jsonc
+{   "result": "error",
+    "error": {
+        "message": "<string>" // an error message, will be the text of the `Error` returned inside `azd`.
+    }
+}
+```
+
+Note that an error prompting leads to a successful result at the HTTP layer (200 OK) but with a special error object. `azd` treats other responses as if the server has an internal bug.
+
+## Open Issues
+
+- [ ] ConsoleOptions today provides a way to provide a function which takes a string and returns an set of suggestions, this is used by our implementation today to tab complete directories.  Do we want to provide similar support here? If so, how might we do that? One could imagine a sort of LRO style thing where the server instead responds with a result like `"completionRequest"` which looks like:
+
+```jsonc
+{
+    "result": "completionRequest",
+    "location": "<url>",
+    "value": "<string>"
+}
+```
+
+And `azd` would execute the completion function, then post the results to the URL listed in `location`, yielding control back
+to the host and allowing it to display the completions.
+
+Another option would be to not support this but instead add a `directory` type that can be prompted for - to allow the host to provide special UI or fall back to a normal string selection for a path, similar to how we handle `password`.
```

Comment:
I think a directory prompt is a good idea. VSCode has a native directory prompting UI that is a lot more user friendly than tab-completing strings, and works better in remote scenarios as well.

Reply:
For reference, https://code.visualstudio.com/api/references/vscode-api#window `showOpenDialog` and [`OpenDialogOptions`](https://code.visualstudio.com/api/references/vscode-api#OpenDialogOptions)

Reply:
I added a "directory" type to the list - I'm not sure if we want this to be `path` and then options for "are files okay, are directories okay?".  Right now `azd` only cares about prompting for a single directory.

Reply:
We could have separate `file` and `directory` types. Files are more involved though, you want to include type filters for optimal experience. IMO we build that bridge when we get to it.